src/posts/2024-02-12-how-software-networking-is-taking-over-the-world/index.md

```diff
@@ -0,0 +1,63 @@
+
+## The Problem With Software
+
+There are many reasons as to why packet processing was originally moved to dedicated networking devices, however, the main and hopefully obvious one is mainly that local (on your server/machine/laptop) packet processing performance would be woeful compared to silicon created solely for that purpose. The speeds of networking interconnects improved at a rapid pace (10mbit -> 100mbit -> 1Gbit -> 10Gbit -> 100 Gbit etc..), and each speed increase would suddenly come with it a large bump up in the amount of data/packets that a CPU would have to process locally. These speeds are usually referred to as “wire-speed” (due to the speeds of traffic on the physical wires), most hardware vendors would aim to ensure that their devices would be capable of processing traffic at “wire-speed” to ensure that the processing work would not incur any bottleneck on the network.
+
+People trying to use commercial hardware/software to replicate the behavior of dedicated silicon would often be met with disappointment. The Journal of Universal Computer Science has a [paper](https://www.jucs.org/jucs_16_21/impact_of_cpu_bound/jucs_16_21_3299_3313_salah.pdf) detailing experiments of replicating high traffic rates from 2010. The journal goes on to state that on a 1 Gigabit connection, an expectation of 700 Kpps (thousand packets per second) is the expected rate. However in testing traffic forwarding on this commercial hardware the resulting rates were around 350 K pps and 75 K pps for both Linux and Windows respectively. Also when reaching these rates the CPU availability for running applications was in the single digits, effectively meaning that we can process the packets but the applications that are receiving the data may well be unresponsive. It’s worth noting that the above experiment was also purely UDP forwarding of packets, in the event we would want to inspect the application data itself and make decisions based upon that we would incur even larger performance penalties.
+
+## Software Catches Up
+
+In the years since that paper was written, there has been a huge evolution in the commercial hardware, CPUs have grown horizontally instead of vertically (single CPU core speed growth has all but stopped, and the focus of chip manufacturers has been multiple packaged cores).
```

Comment:
```suggestion
In the years since that paper was written, there has been a huge evolution in commercial hardware. CPUs have grown horizontally instead of vertically (single CPU core speed growth has all but stopped, and the focus of chip manufacturers has been multiple packaged cores).
```